lib/livebook/runtime/evaluator.ex

```diff
@@ -689,25 +695,169 @@ defmodule Livebook.Runtime.Evaluator do
+
+  # Find the first valid module declaration in a token list
+  defp find_first_module_attribute(tokens) do
+    case :lists.dropwhile(&is_not_module_declaration/1, tokens) do
+      # Multiple declarations of a module could be defined, that's fine, we need at least one to run in module mode
+      [
+        {:-, _},
+        {:atom, _, :module},
+        {:"(", _},
+        {:atom, _, module_name},
+        {:")", _},
+        {:dot, _}
+        | _
+      ] ->
+        {:ok, module_name}
+
+      _ ->
+        # No -module(module_name). attribute
+        :error
+    end
+  end
+
+  # In order to handle the expression in their forms - separate per {:dot,_}
+  defp not_dot({:dot, _}) do
+    false
+  end
+
+  defp not_dot(_) do
+    true
+  end
+
+  # A list of scanned token - must be separated per dot, in order to feed them
+  # into the :erl_parse.parse_form function.
+  defp tokens_to_forms(tokens) do
+    {:ok, tokens_to_forms(tokens, [])}
+  end
+
+  defp tokens_to_forms([], acc) do
+    :lists.reverse(acc)
+  end
+
+  defp tokens_to_forms(tokens, acc) do
+    form = :lists.takewhile(&not_dot/1, tokens)
+    [dot | rest] = :lists.dropwhile(&not_dot/1, tokens)
+    tokens_to_forms(rest, [{form ++ [dot]}] ++ acc)
+  end
+
+  defp parse_forms(form_statements) do
+    try do
+      res =
+        Enum.map(
+          form_statements,
+          fn {form_statement} ->
+            case :erl_parse.parse_form(form_statement) do
+              {:ok, form} -> form
+              err -> throw({:parse_fail, err})
+            end
+          end
+        )
+
+      {:ok, res}
+    catch
+      {:parse_fail, err} -> err
+    end
+  end
+
+  # Create module - tokens from string
+  # Based on: https://stackoverflow.com/questions/2160660/how-to-compile-erlang-code-loaded-into-a-string
+  # The function will first assume that code starting with -module( is a erlang module definition
+
+  # Step 1: Scan the code
+  # Step 2: Convert to forms
+  # Step 3: Extract module name
+  # Step 4: Compile and load
+  # Step 5: If compile success - calculate md5 and register module
+
+  defp eval_module(:erlang, code, tokens, binding, env) do
+    try do
+      with {:ok, form_statements} <- tokens_to_forms(tokens),
+           {:ok, forms} <- parse_forms(form_statements) do
+        # First statement - form = module definition
+        {:attribute, _, :module, module_name} = hd(forms)
+
+        # Compile the forms from the code-block
+        {:ok, _, binary_module} = :compile.forms(forms)
+
+        # Line below should be done in Tracer...
+        {:module, module_name} =
+          :code.load_binary(module_name, ~c"#{module_name}.beam", binary_module)
+
+        # Registration of module
+        # env = %{env | versioned_vars: %{}}
+        env = %{env | module: module_name}
+
+        Evaluator.Tracer.trace({:on_module, binary_module, %{}}, env)
+
+        {{:ok, ~c"erlang module successfully compiled", binding, env}, []}
```

Comment:
Whatever we return is arbitrary, I like `{:ok, module}` :)

```suggestion
        result = {:ok, module_name}
        {{:ok, result, binding, env}, []}
```